### ISSUE TITLE
Proposed content for Major Release 5.0.0

### ISSUE THREAD

**Issue by chrisisla** [finos/common-domain-model#2432]
At the last CRWG #2405 a proposal was made to cut a new production major release of the CDM. The request was put forward by ISDA to allow a production user of the CDM to perform some industry testing of recent changes that have been made to the model.

One of the action points to that meeting was for the other associations and community members to put forward any open Issues or Pull Requests that would need to be included in the proposed release.

Please can you add any critical items to this Issue as soon as possible so that we can determine the content required in the new version and an appropriate timeframe for the release.

@tomhealey-icma @gabriel-ICMA @dshoneisda @rvincentISDA @minesh-s-patel @lolabeis @mgratacos

**Comment by chrisisla** [comment]
From ISLAs perspective the following items are critical for the next release:
- #2161 - required so can qualify a securities lending trade, already in discussion with ICMA on this, hope to have PR for this soon
- #2296 - final modelling for pre-trade processes

There are also some potential changes to the legal agreement structure in the pipeline, I am in discussion with ISDA on these. If changes to the model are required then this would be another item to include the major release.

**Comment by tomhealey-icma** [comment]
We want to include the following PRs in the next prod release:
Add productContractDetails to ProductBase [#2419](https://github.com/finos/common-domain-model/pull/2419)
Updated Repo Partial Delivery Function and Qualification [#2153](https://github.com/finos/common-domain-model/pull/2153)
Updated Repo Reprice [#2121](https://github.com/finos/common-domain-model/pull/2121)
Repo substitution function and qualification [#2046](https://github.com/finos/common-domain-model/pull/2046)

Also we need to resolve issue #2161. Before we can resolve that I've uncovered an issue we need to address. Chris and I had thought the best way to resolve the qualification problem would be to add productDefinitionEnum to productTaxonomy. After attempting this I found that all the qualification functions are based on economicTerms so product level attributes will not work. 

To resolve this we need to consider new qualification process that include product level attributes like productDefinition and legal agreements. Also I noticed that setQualifier has hard coded the source to ISDA. We need to be able to set the source based on the qualification function result.

**Comment by chrisisla** [comment]
@dshoneisda @rvincentISDA 

Please can you provide a list of any open PRs that ISDA would also like to have included in the production release. 

At the moment ISLA and ICMA are working on completing the Issues/PRs that are critical for the new version, but are also being asked to review other ISDA PRs that I do not believe are on the critical path for this release. 

ISLA and ICMA have made considerable concessions about the content of the release in order to expedite it's creation. It would be more efficient for everyone if requests for non-critical reviews/discussions were postponed until version 5.0.0 has been cut. 

Thanks

FYI @tomhealey-icma

**Comment by chrisisla** [comment]
@lolabeis @dshoneisda @tomhealey-icma 

As discussed last Friday, the agreement is to cut the version 5.0.0 major release by or on this Friday 27th October. The following PRs have been raised for consideration/inclusion in the release. I have the following comments on them:

**Add productContractDetails to ProductBase** https://github.com/finos/common-domain-model/pull/2419
This will be resolved in the enhancement to product qualification in PR2161, so this one is essentially being taken care of in a future version.

**Updated Repo Partial Delivery Function and Qualification** https://github.com/finos/common-domain-model/pull/2153
Tom to update so matches current version of code in model, a few changes requested by me to the comments. Once these changes have been made I am happy to approve this change. 

**Updated Repo Reprice** https://github.com/finos/common-domain-model/pull/2121
Tom to fix typos (and new one in Readme too) but again, once the typos have been fixed then I can approve this change too.

**Repo substitution function and qualification** https://github.com/finos/common-domain-model/pull/2046
This change closes the trade for a collateral substitution which is not the best practice as defined in the ISLA and ICMA guidelines. I've added comments to the changes about this. I don't think this PR can be approved in its current format.

**Securities lending Qualification function issue** https://github.com/finos/common-domain-model/issues/2161 
An interim fix has been suggested for this which REGnosys have created some PRs for:

- #2448
- #2456

These need to be reviewed for inclusion in the major release. The strategic fix to qualification has been agreed for inclusion in the next major release, whose release date is TBD.

**Modelling Lender Availability and Borrower Locates** https://github.com/finos/common-domain-model/pull/2296
This is with me to update based upon feedback from Leo and discussion with the ISLA CDM WG. I am hoping to submit these changes today/tomorrow. If I do not get them finalised before Friday then the major release can still be cut, with these changes going into a minor release (i.e. 5.1.0).

Please let me know if I have missed anything.

**Comment by lolabeis** [comment]
Thanks @chrisisla for that summary.

Regarding #2161, we essentially have 2 options, I will respond separately on the PR. We can have a path that gets us there before the 5.0 release while maintaining the target date of 27/10. But as a plan B, if that fix can only go in later, it could get released as a minor (i.e. 5.1.0) as well.

**Comment by lolabeis** [comment]
@tomhealey-icma Can you please kindly approve Chris's PR: #2296 ? I have reviewed the latest and am satisfied with it, so I have approved. However you also requested some changes earlier on, which I believe have been addressed, but it means that your approval is specifically required.

**Comment by dshoneisda** [comment]
@tomhealey-icma We haven't heard anything on your other items. Just to note per the summary above the intention is still to kick off the major release later today/tomorrow. Any items not completed will need to go a minor release.

**Comment by tomhealey-icma** [comment]
I will have the repo PRs submitted by end of day Friday. This is what we agreed.

Also I want to reiterate my concerns that I raised during the meeting:
1. The change to remove repo qualification is short term and temporary. A fix for this should be completed and released before the end of the year which can be in a minor product version release.
2. We need a plan to fix the hard coded source:ISDA.
3. We need a plan to add product qualification with contract information separate from the trade.

**Comment by lolabeis** [comment]
**Update**
* Lender Availability and Borrower Locates (originally #2296, superseded by #2469 ): released in `5.0.0-dev.45`
* Qualification fix (#2456 ): @tomhealey-icma can you please come back on the PR where we proposed 2 options, to confirm the chosen approach? @chrisisla confirmed 1)

**Comment by tomhealey-icma** [comment]
I've replaced the three repo PRs with one new PR, Repo lifecycle additions (Partial Delivery, Reprice, Adjustment, Substitution) [#2472] (https://github.com/finos/common-domain-model/pull/2472). The following PRs are included in the new PRs and will be closed:
Updated Repo Partial Delivery Function and Qualification,https://github.com/finos/common-domain-model/pull/2153
Updated Repo Reprice,https://github.com/finos/common-domain-model/pull/2121
Repo substitution function and qualification,https://github.com/finos/common-domain-model/pull/2046
